lib/logdna/client.rb

```diff
@@ -44,95 +50,125 @@ def process_message(msg, opts = {})
       processed_message
     end
 
-    def schedule_flush
-      start_timer = lambda {
-        sleep(@exception_flag ? @retry_timeout : @flush_interval)
-        flush if @flush_scheduled
-      }
-      Thread.new { start_timer.call }
+    def write_to_buffer(msg, opts)
+      processed_message = process_message(msg, opts)
+      message_size = processed_message.to_s.bytesize
+
+      buffer_byte_size = @lock.synchronize do
```

Comment:
Why not make `buffer` a thread safe array?

Reply:
And `buffer_byte_size` an atom. You pull in `concurrent-ruby` and it has a bunch of thread-safe primitives

Reply:
Maybe. That will only work, I believe, if we can do all the operations with `@buffer` we want atomically in a single step. This mainly concerns the work being done in `#flush_sync`. I think we might be able to rewrite a lot of that in a simple reduce, but need to look closer at the concurrent array.

Reply:
After looking real quick it doesn't seem like arrays support `reduce`. Enumerable does; however, but I don't see an Enumerable class in the concurrent ruby

Reply:
the docs say

> A thread-safe subclass of Array. This version locks against the object itself for every method call, ensuring only one thread can be reading or writing at a time. This includes iteration methods like #each.

so probably worth a deeper look, yeah.

Reply:
There's no way to reduce the operations done to the buffer in a single array function so we can't use that. Additionally the new changes also require locking the flush scheduling so we'd need `@lock` anyways.